### ISSUE TITLE
Problem in preprocessing ALOS2 data with single_scene_stripmapApp.py

### ISSUE THREAD

**Issue by RobinKohrs** [leiyangleon/FSH#7]
I'm trying to preprocess the ALOS2 data now with the `single_scene_stripmapApp`. For that I un-commented  the lines in the file that are specific to ALOS2 and commented the ones for ALOS1. Everything seemed to work and I got the output:

```
Preprocessing Done!
23:27:28
stripmapApp xml Done!
23:27:28
stripmapApp Done!
23:27:28
Single pair of ###test_data/### Done !!!
23:27:28
```

However, I can't find any interferogram folder in my subfolder (that I also called `test_data` like in the example). By what I understood this (or at least one interferogram) should be created with the following line in `single_scene_stripmapApp.py`:

`cmd = 'stripmapApp.py stripmapApp.xml --end=interferogram | tee stripmapApp1.txt' `

I then just tried to execute this line on the command line and got the following error:

`AttributeError: module 'isceobj.StripmapProc' has no attribute 'createRubbersheet'`

I tried to locate this attribute in the `FSH`-code,  but couldn't find anything. Most likely I'm just overlooking something simple. I'm sorry for asking these easy questions. It's just that I'd really like to understand more about using the algorithm properly.

**Comment by RobinKohrs** [comment]
I tried to clean up everything a little bit. And instead of replacing the scripts from `ISCE` I only added the lines in the four files. However, it's still not running. Maybe I should say that in `stripmapApp1.txt` in the `test_data`-folder the content looks like:
```
This is the Open Source version of ISCE.
Some of the workflows depend on a separate licensed package.
To obtain the licensed package, please make a request for ISCE
through the website: https://download.jpl.nasa.gov/ops/request/index.cfm.
Alternatively, if you are a member, or can become a member of WinSAR
you may be able to obtain access to a version of the licensed sofware at
https://winsar.unavco.org/software/isce
Error. The attribute corresponding to the key "dorubbersheeting" is not present in the object "<class '__main__.Insar'>".
Possible causes are the definition in the xml file of such attribute that is no longer defined 
in the object "<class '__main__.Insar'>" or a spelling error
```
And I get the same error now when I run:

`stripmapApp.py stripmapApp.xml --end=interferogram | tee stripmapApp1.txt` 

from the command line

**Comment by leiyangleon** [comment]
The first attached figure indicates there is error in the processing and the stripmap
 processing actually did not start (same start and end time stamp). 

It seems you managed to make the preprocessing script work for ALOS2, which is good. Otherwise, other types of error would have occurred. The error is not related to preprocessing script nor FSH, but related to the missing "rubbersheeting" component in your ISCE installation.

Which version of ISCE are you using? Rubbersheeting is a fairly new step in ISCE v2.2+. Try use the latest version instead.

**Comment by RobinKohrs** [comment]
Thank you very much for the help! I completely uninstalled my conda ISCE-environment and set it up once again. Now when I run `python -c "import isce; print(isce.__version__)"`, it gives me `2.3`
And as I installed it simply from here: https://anaconda.org/conda-forge/isce2, I guess the version is 2.3.3. 

I get the same error when I run the "normal" stripmapmapApp from ISCE on two scenes and add the doRubbersheeting in the xml. 

Is there any chance to fix that by setting up ISCE via Scons?

**Comment by leiyangleon** [comment]
I am not sure if it is going to help. But it is worthwhile to try Scons option instead of conda. 

Another suggestion: since this is completely related to ISCE, why not just post the error message in their user community [forum](http://earthdef.caltech.edu/projects/isce_forum/boards)? I am sure there might have been similar posts and if not, their developers are better aware of the problem than me.

**Comment by RobinKohrs** [comment]
Sure, you are absolutely right! Thanks a lot!

**Comment by RobinKohrs** [comment]
Hi Lei, 

I asked on the ISCE user forum and it seems that in the new `stripmapApp.py` from (ISCE 2.3.3) there is now `doRubbersheetingAzimuth` and `doRubbersheetingRange`. The lines that I inserted from your files however are saying: `doRubbersheeting`. 

I tried to modify the stripmapApp.py and the outputted stripmapAapp.xml from `single_scene_stripmapApp.py`. However I always get some kind of error. So I just wanted to ask if you have any recommendation on which version of ISCE is the best one to use and modify with your code.

**Comment by leiyangleon** [comment]
For the current preprocessing code, we tested with [ISCE v2.2.0](https://winsar.unavco.org/software/isce). Please use it for now. I vaguely remember we also tested with the latest ISCE v2.3. However, I may be wrong. We will try it again later and update the user's guide if necessary.

**Comment by RobinKohrs** [comment]
Thank you very much! Do you know if there is any chance in obtaining the code for version 2.2.0 without being member of one if the institutions listed here: https://winsar.unavco.org/membership/institutions/

**Comment by leiyangleon** [comment]
It is recommended to follow their rules for obtaining the historical versions of ISCE. Although your institution may not be listed there, you can easily file an application as directed on that page to get yours included. 

One more note on the preprocessing code, no matter which ISCE version you use, it is recommended to run the preprocessing code (e.g. `single_scene_stripmapApp.py`) for ALOS-1 dataset first to verify the ISCE installation and preprocessing code setup are successful and etc. Then, you can feel free to test your modified version for ALOS-2 data because the code released here have been designed and verified for ALOS-1 data only. This is useful in case you introduced new errors when editing it for ALOS-2 data.

**Comment by leiyangleon** [comment]
Have you tried to set `doRubbersheetingAzimuth` to true and `doRubbersheetingRange` to false in ISCE v2.3.3? That should mimic the old behavior of only setting `doRubbersheeting` to true based on what their developers said.

**Comment by RobinKohrs** [comment]
Thank you very much for the pointers! Unfortunately there is no ALOS-1 data, at least to my knowledge and by what I see on the asf-site, in our study region. But maybe I'll just try it with some other area to test my installation. 

I tried the `doRubbersheetingAzimuth` alone, and in combination with `doRubbersheetingRange` with the "normal" `stripmapApp` from ISCE on our data, but I get some really strange results.  I will just try it later on again, also by using the output of `single_scene_stripmapApp.py`,  and check if there is something wrong in my xml...

**Comment by leiyangleon** [comment]
**It is always recommended to try the preprocessing code with ALOS-1 data over places where the ALOS data exist, e.g. the test area in the examples we provided on this website.** 

This is important to isolate any possible ISCE installation & setup error from the coding error that might be introduced in your edits for ALOS-2. I understand it is very much desired to directly modify the code and then apply it to ALOS2. But it is strongly recommended to go step by step when dealing with several softwares and codes.

That said, even if you post your strange results, I cannot comment on it and tell you which sources the errors are coming from.

So here is the list of advices:

1. Download the ALOS1 data over our test area in central Maine of United States and process it using single_scene_stripmapApp.py
2. If error, try ISCE's standard stripmapApp command: stripmapApp.py stripmapApp.xml
3. If no error, that means the preprocessing code was not set up properly. Repost it here and I will help you sort that out
4. If still error in Rubbersheeting, set doRubbersheetingAzimuth to True and doRubbersheetingRange to False and run step 2 again
5. If error persists, that confirms the error is associated with your ISCE installation or the way you use it. Post the error messages on the ISCE forum and the developers can help.
6. After the installation/usage error is fixed, re-run step 2 and 1, sequentially to verify and they should both work now.
7. Only when all the above mentioned possible errors/steps have been properly addressed, you can feel free to edit single_scene_stripmapApp.py for ALOS2 and run it over your own data.

**Comment by RobinKohrs** [comment]
Great! Thank you very much !! I'll try to do  it exactly the way you described it with the data over main in the next couple of days. I once used the test data, but only to run `forest_stand_height.py` directly, not the preprocessing.
Once I have the output now I'll post it here. Thanks a lot!:)

**Comment by RobinKohrs** [comment]
Hi Lei, sorry for my late answer. So just to give you some feedback. I really, really appreciate your extensive help and I'm almost a little embarrassed, but I don't manage to run the `single_scene_stripmapApp.py` even with your ALOS-1 data. I downloaded the Scenes: 
`ALPSRP20150821-L1.1` and
`ALPSRP20151002-L1.1`.
That is to say: I manage to run it, but I still get this message in the `stripmapApp1.txt`
```
Error. The attribute corresponding to the key "dorubbersheeting" is not present in the object "<class '__main__.Insar'>".
Possible causes are the definition in the xml file of such attribute that is no longer defined
in the object "<class '__main__.Insar'>" or a spelling error                                                                                   
```
I then manually set `doRubbersheetingAzimuth` to true and `doRubbersheetingRange` to false. Then I ran the `stripmapApp.py` and it sucessfully did the job. However I can't find the necessary files to continue. For example the `topophase.cor.geo` or the `resampOnlyImage.amp.geo`. 

I think I'll setup everything completely new in a VM and try it again.

**Comment by leiyangleon** [comment]
Given these results, now I think that is easy to fix. 

You can modify [format_stripmapApp_xml.py](https://github.com/leiyangleon/FSH/blob/dev/ISCE_processing_scripts/format_stripmapApp_xml.py), and change the line of `rubbersheeting` to two lines of `doRubersheetingAzimuth` (set to True) and `doRubbersheetingRange` (set to False). Then clean the folder with ISCE-created files, and re-run `single_scene_stripmapApp.py`. That should work and let you have the desired results (.cor.geo and .amp.geo).

BTW, the way I set it up the preprocessing code will take ALOS1 L1.0 data as input. Although you can change it to L1.1 with sensor ALOS replaced by ALOS_SLC, you should avoid using L1.1 data for now and exactly mimic what is required for the test examples on this website. Otherwise, you won't be 100% sure in isolating the error sources. You can always change it to ALOS L1.1 or ALOS2 later when you have a working version.

I think by using ALOS L1.0 and following my above tips, you will be all set.

**Comment by RobinKohrs** [comment]
Wow, that will help a lot!! Thank you very much:)! During the week I'll try to do the modifications as you described. However, I fear that there is no Alos-2 Level 1.0 data for our Region. But we'll see what's possible:) Thanks a lot again!

**Comment by leiyangleon** [comment]
ALOS2 only has L1.1 data released.
What I meant was after you have a working version for ALOS1 L1.0 data, you can then modify the preprocessing code for ALOS1 L1.1 data as well as ALOS2 L1.1 data. 
But you need to only worry about the ALOS1 L1.0 data for now and then go step by step as I mentioned earlier.

**Comment by RobinKohrs** [comment]
Hi Lei,
so I downloaded two ALOS1 L1.0 Scenes over Maine. They are the following scenes: 
`ALPSRP077610890-L1.0`
`ALPSRP081840890-L1.0`
I then modified `format_stripmapApp_xml.py`. `single_scene_stripmapApp.py` seems to be working fine. It produces `topophase.cor.geo` (the one in the image). `stripmapProc.xml` is also produced. However, there is no `resampOnlyImage.amp.geo` or  `resampOnlyImage.amp.geo.xml`. 

Probably it's something related to my setup of ISCE. I just don't know where to look for it. I can post the log-files, but maybe I'll just try to set it up once again. Should these two files be in the `interferogram`-directory?
![Screenshot from 2020-06-20 08-14-14](https://user-images.githubusercontent.com/48288925/85194498-15fb0180-b2ce-11ea-92ed-338100bb03ea.png)

**Comment by leiyangleon** [comment]
It seems you might be close, as cor.geo was created successfully. Try to redo the setup as detailed in Section 2.1.1 of the [documentation](https://github.com/leiyangleon/FSH/blob/dev/docs/isce_preprocessing.md).

Another note about your ALOS-1 data: use `ALPSRP091030890-L1.0` for the second scene to reproduce the test example data on this site. The first scene (`ALPSRP077610890-L1.0`) was correct.

If there is still error, please post the `isce.log` file as well as the screenshot of your `interferogram` directory.

**Comment by RobinKohrs** [comment]
Thanks so much! I'll give it another shot:)

**Comment by RobinKohrs** [comment]
Hi Lei, I think I will not manage to get the necessary input files for the `single_scene_stripmapApp.py`. As far as I know the `resampOnlyImage.amp` was only produced in the ISCE  versions until 2.2.0. On the earthdef-forum they said, that now there is the `topophase.amp`-file which is kind of a replacement. But I think it's safe to say that I can't replace one by the other right? And I fear that my skills won't be sufficient to make the necessary adaptions.

**Comment by leiyangleon** [comment]
You don't need to worry about it. You are right that ISCE v2.3+ does not generate those file names in `stripmapApp`. However, that is exactly what our preprocessing script does and why we simplify the preprocessing to be just a single command line (`single_scene_stripmapApp.py`) for the users' convenience. What you need to do is only to set up the preprocessing scripts (by copy&paste) as detailed in the documentation above, and then run that single command line. For using `stripmapApp` in ISCE 2.3+ instead of ISCE 2.2, you only need to modify the rubbersheeting parameter as we have discussed [above](https://github.com/leiyangleon/FSH/issues/7#issuecomment-643686377).

**Comment by RobinKohrs** [comment]
Once again, thank's so much! I tried it again, but didn't succeed. My workflow at the moment looks like this: 
1. Download `ISCE` 2.3.3 via conda
2. Replace the scripts from ISCE with the scripts from `stripmapApp_subsitute` (I also tried not to replace them, but just to insert the specific lines)
3. Change the `format_stripmapApp_xml.py` and set rubbersheetingAzimuth to true and range to false
4. Put the `isce_processing_scripits`-directory on a path that is found by PATH and PYTHONPATH
5. run `single_scene_stripmapApp.py` on the directory with the 2 level-1.0 ALOS-1 data
6. In case this doesn't run directly, execute the produced `stripmapApp.xml` manually with `stripmapApp.py stripmapApp.xml`

At the moment I'm not even able to reproduce the coherence-image from above:/
Would you suggest to simply replace the files from ISCE with all the files in `stripmapApp_substitue` or to look at each line?

**Comment by leiyangleon** [comment]
@RobinKohrs This issue has been resolved in the latest PR #8, where ALOS-2 support and latest ISCE version (v2.4.1) support has been added. Please also see the [release news](https://github.com/leiyangleon/FSH/blob/dev/docs/update.md).
